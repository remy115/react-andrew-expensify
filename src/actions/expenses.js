import uuid from 'uuid';
import {database} from '../firebase/firebase';

// ADD_EXPENSE
export const defaultExpense={
  description:'',
  note:'',
  amount:0,
  createdAt:Date.now()
}
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense=(expense=defaultExpense)=>{
  return (dispatch)=>{
    return database.ref('expenses').push(expense)
      .then((ref)=>{
        return dispatch(addExpense(
          Object.assign({id:ref.key},expense)
        ));
      });
  }
}

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});
