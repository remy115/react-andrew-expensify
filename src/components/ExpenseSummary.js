import React from 'react';
import {connect} from 'react-redux';
import filteredExpenses from '../selectors/expenses';
import getTotalAmount from '../selectors/expenses-total';

export const ExpenseSummaryComp=({expenseCount, expensesTotal})=>{
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
    return (
        <div>
            <p>Viewing {expenseCount} {expenseWord} totalling ${expensesTotal}</p>
        </div>
    );
}

const mapState2Props=(store)=>{
    const filters=store.filters;
    const filteredExpenses1=filteredExpenses(store.expenses,filters);
    return {
        expenseCount:filteredExpenses1.length,
        expensesTotal:getTotalAmount(filteredExpenses1)
    }
}


export default connect(mapState2Props)(ExpenseSummaryComp);