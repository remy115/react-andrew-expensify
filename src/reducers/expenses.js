// Expenses Reducer

const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action={type:'@@INIT'}) => {
  // console.log('@@@@@@@@@@@@@@@',state,action);
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [
        ...state,
        action.expense
      ];
    case 'REMOVE_EXPENSE':
      return state.filter(({ id }) => id !== action.id);
    case 'EDIT_EXPENSE':
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updates
          };
        } else {
          return expense;
        };
      });
    case 'SET_EXPENSES':
      // console.log('################################',Object.assign({},action.expenses));
      return Array.from(action.expenses);
    default:
      return state;
  }
};
