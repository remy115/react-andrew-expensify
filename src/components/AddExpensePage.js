import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

export const AddExpensePage = (props) => (
  <div>
    <h1>Add Expense</h1>
    <ExpenseForm
      onSubmit={(expense) => {
        props.startAddExpense(expense);
        props.history.push('/');
      }}
    />
  </div>
);

const mapDispatch2Props=(dispatch)=>{
  return {
    startAddExpense:(expense)=>dispatch(startAddExpense(expense))
  }
}

export default connect(undefined,mapDispatch2Props)(AddExpensePage);
