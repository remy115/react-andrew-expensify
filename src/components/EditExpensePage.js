import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export const EditExpensePage = (props) => {
  return (
    <div>
      <div className="page-header">
        <div className="content-container">
          <h1 className="page-header__title">Edit Expense</h1>
        </div>
      </div>
      <div className="content-container">
        <ExpenseForm
          expense={props.expense}
          onSubmit={(expense) => {
            props.startEditExpense(props.expense.id, expense)
              .then(()=>{
                props.history.push('/');
              });
          }}
        />
        <button className="button button--secondary" onClick={() => {
          props.startRemoveExpense({ id: props.expense.id })
            .then(()=>{

              props.history.push('/');
            });
        }}>Remove</button>
      </div>
    </div>
  );
};

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  };
};

const mapDispatch2Props=(dispatch)=>{
  return {
    startEditExpense:(id,expense)=>dispatch(startEditExpense(id, expense)),
    startRemoveExpense:(objId)=>dispatch(startRemoveExpense(objId))
  }
}

export default connect(mapStateToProps,mapDispatch2Props)(EditExpensePage);
