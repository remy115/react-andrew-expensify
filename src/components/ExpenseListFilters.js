import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';


export class ExpenseListFilters extends React.Component {
  constructor(props) {
    super(props);
    this.state={
      calendarFocused: null

    }
    this.onDatesChange=this.onDatesChange.bind(this);
    this.onFocusChange=this.onFocusChange.bind(this);

    // console.log('props.filters',props.filters);
  }

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  }
  render() {
    return (
      <div>
        <input
          type="text"
          name="text"
          value={this.props.filters.text}
          onChange={(e) => {
            this.props.setTextFilter(e.target.value);
          }}
        />
        <select
          value={this.props.filters.sortBy}
          onChange={(e) => {
            if (e.target.value === 'date') {
              this.props.sortByDate();
            } else if (e.target.value === 'amount') {
              this.props.sortByAmount();
            }
          }}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          startDateId="start"
          endDate={this.props.filters.endDate}
          endDateId="end"
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          showClearDates={true}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
      </div>
    );
  }
};

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  };
};

const mapDispatch2Props=(dispatch)=>({
  setStartDate:(startDate)=>dispatch(setStartDate(startDate)),
  setEndDate:(endDate)=>dispatch(setEndDate(endDate)),
  setTextFilter:(text)=>dispatch(setTextFilter(text)),
  sortByDate:()=>dispatch(sortByDate()),
  sortByAmount:()=>dispatch(sortByAmount())
})

export default connect(mapStateToProps,mapDispatch2Props)(ExpenseListFilters);
