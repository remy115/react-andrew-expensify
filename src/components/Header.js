import React from 'react';
import { Link } from 'react-router-dom';
import {startLogout} from '../actions/auth';
import {connect} from 'react-redux';

export const Header = ({startLogout}) => (
  <header className="header">
    <div className="content-container">
      <div className="header__content">
        <Link className="header__title" to="/dashboard" >
          <h1>Expensify</h1>
        </Link>
        <Link to="/create" activeClassName="is-active">Create Expense</Link>
        <button type="button" onClick={startLogout}>Logout</button>
      </div>
    </div>
  </header>
);

const mapDispatch2Props=(dispatch)=>{
  return {
    startLogout:()=>dispatch(startLogout())
  }
}

export default connect(null,mapDispatch2Props)(Header);