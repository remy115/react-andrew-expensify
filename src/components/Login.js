import React from 'react';
import {connect} from 'react-redux';
import {startLogin} from '../actions/auth';

export class Login extends React.Component {
    constructor(props) {
        super(props);
        
        this.state={
            login:'',
            pass:''
        }        
        this.onSubmit=this.onSubmit.bind(this);
    }

    onSubmit(e) {
        e.preventDefault();
        if(!this.state.login || !this.state.pass) {
            return alert('Fill login and password!');
        }
    }

    onChange(e) {
        const name=e.target.name;
        const value=e.target.value;

        this.setState(()=>({
            [name]:value
        }))))));
    }

    render() {
        return (
            <div className="box-layout" >
                <div className="box-layout__box">
                    <h2 className="box-layout__title">Expensify</h2>
                    <p>It's time to get your expenses under control</p>
                    <button type="button" className="button" onClick={this.props.startLogin}>Login with Google</button>
                </div>
            </div>
        )
    }
}

const mapDispatch2Props=(dispatch)=>{
    return {
        startLogin:()=>dispatch(startLogin())
    }
}

export default connect(null,mapDispatch2Props)(Login);