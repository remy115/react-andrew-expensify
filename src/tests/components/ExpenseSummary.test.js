import React from 'react';
import {shallow} from 'enzyme';
import {ExpenseSummaryComp} from '../../components/ExpenseSummary';
import getTotalAmount from '../../selectors/expenses-total';

import expenses from '../fixtures/expenses';

let wrapper;
beforeEach(()=>{
    wrapper=shallow(<ExpenseSummaryComp expenseCount={expenses.length} expensesTotal={getTotalAmount(expenses)} />);
});

it('should render ExpenseSummary with more than 1 expense',()=>{
    expect(wrapper).toMatchSnapshot();
});


it('should render ExpenseSummary with just one expense',()=>{
    const wrapper=shallow(<ExpenseSummaryComp expenses={[expenses[1]].length} total={getTotalAmount([expenses[1]])} />);
    expect(wrapper).toMatchSnapshot();
});