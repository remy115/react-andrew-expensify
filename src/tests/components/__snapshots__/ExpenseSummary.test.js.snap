// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseSummary with just one expense 1`] = `
<div>
  <p>
    Viewing 
     
    expenses
     totalling $
  </p>
</div>
`;

exports[`should render ExpenseSummary with more than 1 expense 1`] = `
<div>
  <p>
    Viewing 
    3
     
    expenses
     totalling $
    1141.95
  </p>
</div>
`;
